### PR TITLE
Authentication APIs for IBM PFW SMS Menu

### PR DESCRIPTION
**Pull request by joseph-reynolds**
Background: The IBM Partition firmware SMS component (external to the BMC) has a menu which requires authentication (username and password) to access.  It uses the BMC as its authentication server via PLDM transport.

This implements IBM PFW SMS Menu authentication and change password APIs. These APIs use the BMC's Linux-PAM service to access BMC user data.

Specifically:
 - A new header file declares the new functions invoked by the PLDM service.
 - A new C++ module implements the new functions.
 - A new unit test tool directly invokes the new functions.
 - A new Linux-PAM config file is provided.

Tested: via the BMC command shell with root authority First create a user account
try-pfw-sms-menu test bad  # verify auth failed
useradd test
current_password="0penBmc4Everyone"
echo $current_password | sed p | passwd test

Test authentication scenarios
try-pfw-sms-menu test $current_password
verify authentication was successful, not authorized 
usermod --append -G priv-admin test
try-pfw-sms-menu test $current_password  # authenticated, authorized 
passwd --expire test
try-pfw-sms-menu test $current_password  # password change required

Test password change scenarios
try-pfw-sms-menu test $current_password abcdefg  # too short 
try-pfw-sms-menu test $current_password abcdefgh  # <2 character classes 
try-pfw-sms-menu test $current_password 'ABCdefgh'  # monotonic sequence >3 
try-pfw-sms-menu test $current_password 'r@nd test BMC!'  # has username 
try-pfw-sms-menu test $current_password ''  # Zero length password 
try-pfw-sms-menu test $current_password 'gHw%%wHg'  # palindrome 
try-pfw-sms-menu test $current_password '1password'  # dictionary word

usermod --append -G ipmi test   # Hack for unit testing, do not use
try-pfw-sms-menu test $current_password 'IS-too-long-for-IPMI!'

### CODE REVIEW

File: oem/ibm/pfw-sms/meson.build [hunk 1]
```diff
@@ -0,0 +1,12 @@
+libpam = cpp.find_library('pam', required: true)
+try_pfw_sms_deps = [libpam]
+
+executable(
+  'try-pfw-sms-menu',
+  'try_pfw_sms_ibm.c',
+  'pfw_sms_menu_ibm.c',
```
Comment (jaypadath) [inline oem/ibm/pfw-sms/meson.build:7]: Please change the .c extension to .cpp and .hpp. Also the folder structure is under ibm you may omit it from the file name itself.

File: meson.build [hunk 2]
```diff
@@ -232,6 +232,11 @@ subdir('pldmtool')
 
 subdir('configurations')
 
+if get_option('oem-ibm').enabled()
+  subdir('config')
+  subdir('oem/ibm/pfw-sms')
```
Comment (jaypadath) [inline meson.build:237]: would be good to name the folder as pfw-sms to pfw-sms-utils

File: libpldmresponder/meson.build [hunk 3]
```diff
@@ -80,9 +80,15 @@ if get_option('oem-ibm').enabled()
     '../oem/ibm/libpldmresponder/file_io_type_vpd.cpp',
     '../oem/ibm/host-bmc/host_lamp_test.cpp',
     '../oem/ibm/libpldmresponder/bios_oem_ibm.cpp',
+    '../oem/ibm/pfw-sms/pfw_sms_menu_ibm.c',
```
Comment (jaypadath) [inline libpldmresponder/meson.build:83]: Please change the .c extension to .cpp and .hpp. Also the folder structure is under ibm you may omit it from the file name itself.

File: oem/ibm/pfw-sms/try_pfw_sms_ibm.c [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// unit test stub
```
Comment (jaypadath) [inline oem/ibm/pfw-sms/try_pfw_sms_ibm.c:1]: We need to go with the Gtest framework here

**Review by jaypadath — CHANGES_REQUESTED** [review]

File: oem/ibm/pfw-sms/pfw_sms_menu_ibm.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * @param[in] password - the password associated with the username
+ * @param[out] authenticated - true if password authentication was successful; note this only reports password authentication, and the parameters passwordChangeRequired and operationAllowed should be checked as appropriate.
+ * @param[out] passwordChangeRequired - true if the password needs to be changed
+ * @param[out] operationAllowed - true if the user should be allowed to perform the PFW SMS operation
+ *
+ * @note Authentication is to BMC user account
+*/
+void authenticate(const char *username, const char *password, bool *authenticated, bool *passwordChangeRequired, bool *operationAllowed);
+
+enum changePasswordReasonCode
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms/pfw_sms_menu_ibm.h:24]: enum could be moved first

File: oem/ibm/pfw-sms/pfw_sms_menu_ibm.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+/** @brief Handle password change request from PFW SMS
+ *
+ * @param[in] username - the name of the BMC account
+ * @param[in] currentPassword - the password associated with the username
+ * @param[in] newPassword - the proposed new password
+ * @return changePasswordReasonCode
+ */
+enum changePasswordReasonCode
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms/pfw_sms_menu_ibm.h:51]: Please create class based model which handles all the functionalities and please use Gtest framework for unit test as suggested by @jaypadath .

File: oem/ibm/pfw-sms/meson.build (hunk 1 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms/meson.build:7]: Okay, I had previously understood my code should be in C (not C++).  I will make this change.

File: meson.build (hunk 2 above)
Comment (joseph-reynolds) [reply meson.build:237]: Okay

File: libpldmresponder/meson.build (hunk 3 above)
Comment (joseph-reynolds) [reply libpldmresponder/meson.build:83]: Okay, I will do this.

File: oem/ibm/pfw-sms/pfw_sms_menu_ibm.h (hunk 5 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms/pfw_sms_menu_ibm.h:24]: Okay, will do.

File: oem/ibm/pfw-sms/pfw_sms_menu_ibm.h (hunk 6 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms/pfw_sms_menu_ibm.h:51]: I could do this, but I don't know how that would help.  The class would have only 2 member functions and the enum declaration.  And the difficulties I found trying to use gtest on this are outlined in a comment above.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                  bool* authenticated, bool* passwordChangeRequired,
+                  bool* operationAllowed)
+{
+    common_authentication_handler(username, password, authenticated,
+                                  passwordChangeRequired, operationAllowed);
+}
+
+enum changePasswordReasonCode changePassword(const char* username,
+                                             const char* currentPassword,
+                                             const char* newPassword)
```
Comment (jaypadath) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:233]: Missing operationAllowed flag

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+     * as appropriate.
+     * @param[out] passwordChangeRequired - true if the password needs to be
+     * changed
+     * @param[out] operationAllowed - true if the user should be allowed to
+     * perform the PFW SMS operation
+     *
+     * @note Authentication is to BMC user account
+     */
+    void authenticate(const char* username, const char* password,
+                      bool* authenticated, bool* passwordChangeRequired,
```
Comment (jaypadath) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp:57]: Change from * to &

**Review by jaypadath — CHANGES_REQUESTED** [review]

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+{
+#endif
+
+#include <security/pam_appl.h>
+#include <stdbool.h>
+#include <string.h>
+
+    enum changePasswordReasonCode
+    {
+        PasswordChangeSuccessful = 0,
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp:15]: can you please use enum name as per other pldm supported enum?
[https://github.com/ibm-openbmc/libpldm/blob/efb400697eede42f14ad6290660a4aecbc314141/include/libpldm/base.h#L20](url)

Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp:15]: Yes, I will do that.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp (hunk 8 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp:57]: Yes, I was doing this as part of converting form C to C++.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        return;                     // *operationAllowed = false
+    }
+    *passwordChangeRequired = (retval == PAM_NEW_AUTHTOK_REQD);
+    *operationAllowed = (!*passwordChangeRequired);
+
+    retval = pam_end(pamHandle, retval); // ignore retval
+}
+
+void authenticate(const char* username, const char* password,
+                  bool* authenticated, bool* passwordChangeRequired,
```
[the remark this answers is not in the document]
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:224]: Yes, I am doing that as part of converting this from C to C++.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        }
+    }
+    return PAM_SUCCESS;
+}
+
+/** @brief Common code factored out between authenticate() and changePassword().
+ */
+static void common_authentication_handler(const char* username,
+                                          const char* password,
+                                          bool* authenticated,
```
[the remark this answers is not in the document]
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:180]: Yes, I will do that as part of converting this from C to C++.

**Comment by joseph-reynolds** [comment]
I am still working on adding gtest tests.

**Comment by joseph-reynolds** [comment]
I see the CI checks have failed, but I am unable to see them.  Jenkins reports: _This page may not exist, or you may not have permission to see it._

**Comment by joseph-reynolds** [comment]
Note when unit testing this.  These functions use library function pam_authenticate() which needs to the `/etc/pam.d/ibm-pfw-sms-menu` file present, or else it will always return an authentication error (code 7).

I've updated the meson.build files to copy the new `ibm-pfw-sms-menu` _PAM-configuration_ file into the right directory.  If this is wrong, please let me know the proper way to do this.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+/** @brief Return code from changePassword
+ *
+ * The "should never happen" codes should be captured for debug/diagnosis.
+ * Comments describe the conditions for which the other codes are used;
+ * these can be given to interactive users.
+ */
+enum changePasswordReasonCode
+{
+    PASSWORDCHANGESUCCESSFUL = 0x00,
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp:17]: Could you please use same format as PAM_CONV_ERR?
adding underscore between words, it would help to align with other enum parameter and easy to read.

**Comment by joseph-reynolds** [comment]
jenkins run tests please

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp (hunk 12 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.hpp]: Yes, done in latest patchset

File: oem/ibm/pfw-sms-utils/try_pfw_sms_ibm.cpp [hunk 13]
```diff
@@ -0,0 +1,43 @@
+// unit test stub
```
Comment (jaypadath) [inline oem/ibm/pfw-sms-utils/try_pfw_sms_ibm.cpp:1]: Please remove this file as this can be enabled later with Gtest frame work

File: oem/ibm/pfw-sms-utils/meson.build [hunk 14]
```diff
@@ -0,0 +1,14 @@
+libpam = cpp.find_library('pam', required: true)
+try_pfw_sms_deps = [libpam]
+
+executable(
+  'try-pfw-sms-menu',
+  'try_pfw_sms_ibm.cpp',
```
Comment (jaypadath) [inline oem/ibm/pfw-sms-utils/meson.build:6]: Please remove these try files and dependencies as these can be enabled with Gtest in another commit

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+    }
+    const struct pam_message* msg_ptr = *message_ptr;
+    for (int msg_index = 0; msg_index < num_msg; ++msg_index)
+    {
+        char* local_password = NULL; // for PAM's malloc'd copy
+        std::basic_string_view msg(msg_ptr[msg_index].msg);
+        switch (msg_ptr[msg_index].msg_style)
+        {
+            case PAM_PROMPT_ECHO_OFF:
+                // std::cout << "PAM prompt echo off: " << msg << "\n";
```
Comment (jaypadath) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:75]: Please remove all the std::couts. We do use LG2 logging in pldm. If you want to trace any con-confidential info, include below file:
#include <phosphor-logging/lg2.hpp>

Example trace:
info("common_authentication_handler pam_authenticate failed retval={AUTH_RETVAL}", "AUTH_RETVAL", retval)

**Review by jaypadath — CHANGES_REQUESTED** [review]

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 15 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:75]: Okay, I will remove them; they were for unit testing.  I don't need these for diagnostic.

File: oem/ibm/pfw-sms-utils/meson.build (hunk 14 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/meson.build:6]: Yes, I will remove this.

File: oem/ibm/pfw-sms-utils/try_pfw_sms_ibm.cpp (hunk 13 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/try_pfw_sms_ibm.cpp:1]: I will remove this.  And plan to replace with gtest.

File: meson.build [hunk 16]
```diff
@@ -232,6 +232,10 @@ subdir('pldmtool')
 
 subdir('configurations')
 
+if get_option('oem-ibm').enabled()
+  subdir('oem/ibm/pfw-sms-utils')
```
Comment (manojkiraneda) [inline meson.build:236]: can we create this folder under oem/ibm/libpldmresponder ?

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+ * function's appdata_ptr) to pass data to and from our application and PAM:
+ * @note For parameter descriptions, see docs for pam_start and pam_conv
+ * @return PAM status code (int)
+ */
+static int pamConversationFunction(int num_msg,
+                                   const struct pam_message** message_ptr,
+                                   struct pam_response** response_ptr,
+                                   void* appdata_ptr)
+{
+    if ((appdata_ptr == NULL) || (message_ptr == NULL) ||
```
Comment (manojkiraneda) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:47]: replace all NULL with nullptr everywhere.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+    {
+        return PAM_CONV_ERR;
+    }
+    if (num_msg <= 0 || num_msg >= PAM_MAX_NUM_MSG)
+    {
+        return PAM_CONV_ERR;
+    }
+
+    struct pfw_sms_pam_appdata* pfw_sms_appdata_ptr =
+        (struct pfw_sms_pam_appdata*)appdata_ptr;
```
Comment (manojkiraneda) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:58]: avoid c style casting , prefer c++ based static or re-interpret-cast

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+    *response_ptr = (pam_response*)calloc(
+        num_msg, sizeof(struct pam_response)); // PAM frees
+    if (*response_ptr == NULL)
+    {
+        return PAM_CONV_ERR;
+    }
+    const struct pam_message* msg_ptr = *message_ptr;
+    for (int msg_index = 0; msg_index < num_msg; ++msg_index)
```
Comment (manojkiraneda) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:66]: range based for loop

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+    {
+        char* local_password = NULL; // for PAM's malloc'd copy
```
Comment (manojkiraneda) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:68]: nullptr

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+        std::basic_string_view msg(msg_ptr[msg_index].msg);
+        switch (msg_ptr[msg_index].msg_style)
+        {
+            case PAM_PROMPT_ECHO_OFF:
+                // Assume PAM is asking for the password.  Supply a malloc'd
+                // password that PAM will free
+                local_password = ::strdup(
```
Comment (manojkiraneda) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:75]: please use the std::string class , no c-string manipulation API's.

**Review by manojkiraneda — CHANGES_REQUESTED** [review]

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 18 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:58]: Will change to use reinterpret_cast

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 17 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:47]: Will do.  This started as C code; will change to nullptr.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 20 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:68]: Will change

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 21 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:75]: Sorry, I cannot do that.  The Linux-PAM code requires my code to malloc() (or calloc()) and it performs free(), anbd I cannot change Linux-PAM, so we are locked into this old-style memory management.  I added a comment in several places to explain why we must do this.

File: meson.build (hunk 16 above)
Comment (manojkiraneda) [reply meson.build:236]: thinking again , i think we are okay with oem/ibm/pfw-sms-utils , as its a util API and does not really not a util libpldmresponder API. please ignore my comment about movement.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 21 above)
Comment (manojkiraneda) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:75]: ah okay , then having comments to make people understand reason is good enough. Thanks joseph.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+        return PAM_CONV_ERR;
+    }
+    if (num_msg <= 0 || num_msg >= PAM_MAX_NUM_MSG)
+    {
+        return PAM_CONV_ERR;
+    }
+
+    struct pfw_sms_pam_appdata* pfw_sms_appdata_ptr =
+        reinterpret_cast<struct pfw_sms_pam_appdata*>(appdata_ptr);
+    // Malloc array of responses that PAM will free; required to use malloc.
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:60]: comment is misleading from malloc v/s calloc.
comment has malloc but actual code uses calloc.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+                                   void* appdata_ptr)
+{
+    if ((appdata_ptr == nullptr) || (message_ptr == nullptr) ||
+        (response_ptr == nullptr))
+    {
+        return PAM_CONV_ERR;
+    }
+    if (num_msg <= 0 || num_msg >= PAM_MAX_NUM_MSG)
+    {
+        return PAM_CONV_ERR;
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:55]: could you please  improve tracing in case of failure cases.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+
+    struct pfw_sms_pam_appdata appdata = {
+        password.c_str(), PASSWORD_CHANGE_FAILED_UNKNOWN_REASON};
+    char* appData = reinterpret_cast<char*>(&appdata);
+    const struct pam_conv pamConversation = {pamConversationFunction, appData};
+    pam_handle_t* pamHandle = nullptr; // this gets set by pam_start
+    int retval = pam_start("ibm-pfw-sms-menu", username.c_str(),
+                           &pamConversation, &pamHandle);
+    if (retval != PAM_SUCCESS)
+    {
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:190]: is pam_end missing here??
pam_end(pamHandle, retval); // ignore retval

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 25]
```diff
@@ -0,0 +1,280 @@
...
+    }
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        return;                     // operationAllowed == false
+    }
+    operationAllowed = true;
+    passwordChangeAllowed = true;
+
+    retval = pam_end(pamHandle, retval); // ignore retval
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:222]: assigning of retval won't make any sense since it's last line and retval is local scope & function returns nothing.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+
+    retval = pam_chauthtok(pamHandle, PAM_SILENT);
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        // Return the specific reason gathered from the conversation function
+        return appdata.reasonCode;
+    }
+
+    retval = pam_end(pamHandle, retval); // ignore retval
```
Comment (Patel-Kamalkumar) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:276]: assigning of retval won't make any sense since it's last line and retval is local scope & function returns nothing.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 24 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:190]: No.  I will add a comment to indicate that pam_start() failed, so no cleanup is needed.

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 26 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:276]: Same comment as above (at a different place in the code).  I will remove the assignment.

And note it is customary practice for Linux-PAM clients to ignore the return value from pam_end() when the previous operations have all been successful.  For example: https://github.com/linux-pam/linux-pam/blob/master/examples/vpass.c

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp (hunk 23 above)
Comment (joseph-reynolds) [reply oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:55]: Individual PAM modules (such as pam_pwquality.so) write messages to the journal; this is controlled by the /etc/pam.d/ files, using the QUIET argument.
I don't have any ideas for what more to trace at this layer.  In this particular line of code, num_msg==1 always, for historical reasons, so we are already being paranoid here, and never expect this error.

**Review by Patel-Kamalkumar — APPROVED** [review]
Looks good to me, @manojkiraneda and @jaypadath need to look at once.

**Review by jaypadath — APPROVED** [review]

File: oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp [hunk 27]
```diff
@@ -55,26 +56,27 @@ static int pamConversationFunction(int num_msg,
...
+    // Malloc array of responses that PAM will free; required to use malloc.
+    *response_ptr = reinterpret_cast<struct pam_response*>(calloc(
+        num_msg, sizeof(struct pam_response))); // PAM will free
+    if (*response_ptr == nullptr)
     {
         return PAM_CONV_ERR;
     }
     const struct pam_message* msg_ptr = *message_ptr;
-    for (int msg_index = 0; msg_index < num_msg; ++msg_index)
+    for (const int msg_index : std::ranges::iota_view{1, num_msg})
```
Comment (joseph-reynolds) [inline oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp:69]: This change introduced a bug which is fixed by PR 523.  The range should start from zero.
